src/System.Security.Cryptography.Pkcs/tests/SignedCms/SignedCmsTests.netcoreapp.cs

```diff
@@ -0,0 +1,498 @@
+
+            Func<(X509Certificate2, AsymmetricAlgorithm)> getFirstSignerCertAndKey = GetECDsaCertAndKey;
+            Func<AsymmetricAlgorithm, bool> signHashCalled = (key) => ((MyECDsa)key).SignHashCalled;
+
+            Func<(X509Certificate2, AsymmetricAlgorithm)> getCounterSignerCertAndKey = GetRSACertAndKey;
+            Func<AsymmetricAlgorithm, bool> counterSignHashCalled = (key) => ((MyRSA)key).SignHashCalled;
+
+            (byte[] signedData, X509Certificate2 signerCert, X509Certificate2 counterSignerCert) = GetCounterSignedData(content, getFirstSignerCertAndKey, getCounterSignerCertAndKey, signHashCalled, counterSignHashCalled);
+
+            using (signerCert)
+            using (counterSignerCert)
+            {
+                VerifyCmsCorrectlySigned(content, signedData, signerCert, counterSignerCert);
+            }
+        }
+
+        [Fact]
+        public static void SignCmsUsingRSACertAndECDsaKeyThrows()
+        {
+            byte[] content = { 9, 8, 7, 6, 5 };
+
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            using (X509Certificate2 cert = Certificates.RSA2048SignatureOnly.GetCertificate())
+            using (ECDsa key = ECDsa.Create())
+            {
+                CmsSigner signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, cert, key);
+                Assert.Throws<CryptographicException>(() => cms.ComputeSignature(signer));
+            }
+        }
+
+        [Fact]
+        public static void SignCmsUsingDSACertAndECDsaKeyThrows()
+        {
+            byte[] content = { 9, 8, 7, 6, 5 };
+
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            using (X509Certificate2 cert = Certificates.Dsa1024.GetCertificate())
+            using (ECDsa key = ECDsa.Create())
+            {
+                CmsSigner signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, cert, key);
+                signer.IncludeOption = X509IncludeOption.EndCertOnly;
+                signer.DigestAlgorithm = new Oid(Oids.Sha1, Oids.Sha1);
+                Assert.Throws<CryptographicException>(() => cms.ComputeSignature(signer));
+            }
+        }
+
+        [Fact]
+        public static void SignCmsUsingEDCSaCertAndRSAaKeyThrows()
+        {
+            byte[] content = { 9, 8, 7, 6, 5 };
+
+            ContentInfo contentInfo = new ContentInfo(content);
+            SignedCms cms = new SignedCms(contentInfo, detached: false);
+
+            using (X509Certificate2 cert = Certificates.ECDsaP256Win.GetCertificate())
+            using (RSA key = RSA.Create())
+            {
+                CmsSigner signer = new CmsSigner(SubjectIdentifierType.SubjectKeyIdentifier, cert, key);
+                Assert.Throws<CryptographicException>(() => cms.ComputeSignature(signer));
```

Comment:
If you wanted to share code with the three non-matching-key tests it looks like you just need to pass in a Certificate object (to call `GetCertificate()` on) and a key.  But it's small enough that it doesn't seem necessary.